source/write/gridfs.txt

```diff
@@ -0,0 +1,265 @@
+the files when uploaded to a bucket.
+
+.. figure:: /includes/figures/GridFS-upload.png
+   :alt: A diagram that shows how GridFS uploads a file to a bucket
+
+When retrieving files, GridFS fetches the metadata from the ``files``
+collection in the specified bucket and uses the information to reconstruct
+the file from documents in the ``chunks`` collection. You can read the file
+into memory or output it to a stream.
+
+.. _gridfs-create-bucket:
+
+Create a GridFS Bucket
+----------------------
+
+To store or retrieve files from GridFS, create a GridFS bucket by calling the
+``GridFSBucket()``constructor and passing in a ``Database`` instance.
+You can use the ``GridFSBucket`` instance to
+call read and write operations on the files in your bucket.
+
+.. literalinclude:: /includes/gridfs/gridfs.py
+      :language: python
+      :copyable: true
+      :start-after: start create bucket
+      :end-before: end create bucket
+
+.. _gridfs-create-custom-bucket:
+
+To create or reference a bucket with a custom name other than the default name
+``fs``, pass your bucket name as the second parameter to the ``GridFSBucket()``
+constructor, as shown below:
+
+.. literalinclude:: /includes/gridfs/gridfs.py
+      :language: python
+      :copyable: true
+      :start-after: start create custom bucket
+      :end-before: end create custom bucket
+
+.. _gridfs-upload-files:
+
+Upload Files
+------------
+
+Use the ``open_upload_stream()`` method from the ``GridFSBucket`` class to create an upload
+stream for a given file name. The
+``open_upload_stream()`` method allows you to specify configuration information
+such as file chunk size and other field/value pairs to store as metadata. Set
+these options as parameters of ``open_upload_stream()``, as shown in the
+following code example:
+
+.. literalinclude:: /includes/gridfs/gridfs.py
+      :language: python
+      :copyable: true
+      :start-after: start upload files
+      :end-before: end upload files
+
+.. _gridfs-retrieve-file-info:
+
+Retrieve File Information
+-------------------------
+
+In this section, you can learn how to retrieve file metadata stored in the
+``files`` collection of the GridFS bucket. The metadata contains information
+about the file it refers to, including:
+
+- The ``_id`` of the file
+- The name of the file
+- The length/size of the file
+- The upload date and time
+- A ``metadata`` document in which you can store any other information
+
+To retrieve files from a GridFS bucket, call the ``find()`` method on the ``GridFSBucket``
+instance. The method returns a ``Cursor`` instance
+from which you can access the results.
+
+The following code example shows you how to retrieve and print file metadata
+from all your files in a GridFS bucket. Among the different ways that you can
+traverse the retrieved results from the ``Cursor`` iterable, the
+following example uses the ``for...of`` syntax to display the results:
```

Comment:
S: Could link to the Cursors page her for more information. :ref:`pymongo-cursors`